Sources/SwiftKafka/KafkaPollingSystem.swift

```diff
@@ -0,0 +1,459 @@
+
+extension KafkaPollingSystem {
+    /// The state machine used by the ``KafkaPollingSystem``.
+    struct StateMachine {
+        /// The possible states of the state machine.
+        enum State {
+            /// The state machine has been initialized with init() but is not yet Initialized
+            /// using `func initialize()` (required).
+            case uninitialized
+            /// Initialized state (idle).
+            case idle(
+                source: Producer.Source?,
+                pollClosure: (() -> Void)?,
+                running: Bool
+            )
+            /// The system is up and producing acknowledgement messages.
+            case producing(
+                source: Producer.Source?,
+                pollClosure: (() -> Void)?,
+                running: Bool
+            )
+            /// The poll loop is currently suspended and we are waiting for an invocation
+            /// of `produceMore()` to continue producing messages.
+            case stopProducing(
+                source: Producer.Source?,
+                continuation: CheckedContinuation<Void, Error>?,
+                pollClosure: (() -> Void)?,
+                running: Bool
+            )
+            /// The system is shut down.
+            case finished
+        }
+
+        /// The current state of the state machine.
+        private var state: State = .uninitialized
+
+        /// Delayed initialization of `StateMachine` as the `source` and the `pollClosure` are
+        /// not yet available when the normal initialization occurs.
+        mutating func initialize(
+            source: Producer.Source,
+            pollClosure: @escaping () -> Void
+        ) {
+            guard case .uninitialized = self.state else {
+                fatalError("\(#function) can only be invoked in state .uninitialized, but was invoked in state \(self.state)")
+            }
+            self.state = .idle(source: source, pollClosure: pollClosure, running: false)
+        }
+
+        /// Actions to take after ``run()`` has been invoked on the ``KafkaPollingSystem/StateMachine``.
+        enum RunAction {
+            /// Do nothing, the poll loop is already terminated.
+            case alreadyClosed
+            /// The poll loop is already running. ``run()`` should not be invoked more than once!
+            case alreadyRunning
+            /// The poll loop can now be started.
+            case startLoop
+        }
+
+        mutating func run() -> RunAction {
+            switch self.state {
+            case .uninitialized:
+                fatalError("\(#function) invoked while still in state .uninitialized")
+            case .idle(let source, let pollClosure, let running):
+                guard running == false else {
+                    return .alreadyRunning
+                }
+                self.state = .idle(source: source, pollClosure: pollClosure, running: true)
+            case .producing(let source, let pollClosure, let running):
```

Comment:
Can this happen? How can we be in `producing` before `run` was called?

Reply:
I thought this might be possible e.g. when `produceMore()` gets called upon the `NIOAsyncSequenceProducerDelegate` before `run()` gets invoked

Reply:
That won't happen. `produceMore()` only gets called after `yield()` returns `stopProducing`